### ISSUE TITLE
RFC Separate WebAlerts, WebBaskets, and query logging

### ISSUE THREAD

**Issue by jirikuncar** [inveniosoftware/invenio#2622]
* [ ] Allow user to create alerts based on any query. Alert engine MUST respect user id of creator.
* [ ] Allow creating alerts with different notification backends - email, tags/baskets, web notification.
* [ ] Remove support for `p1`, `f1`, `p2`, `f2`, etc. and use only `p` (`query`) argument (this might be touch also OAI Server).
* [ ] Add notification system that can be used in webhooks (cc @lnielsen-cern)

... More details will come together with #2587 and #2619.

**NOTE** Can we go for fast track with this RFC because it is kind of blocker for new search engine using *Invenio-Query-Parser*.

--
@egabancho @kasioumis @kaplun @tiborsimko @zenodo

**Comment by tiborsimko** [comment]
:+1: 

As discussed IRL, we don't have any active "labs" use case for alerts or baskets on Invenio v2.x yet, so this separation would make sense also from the further modularisation and implementation time plan point of view.

**Comment by jirikuncar** [comment]
> we don't have any active "labs" use case for alerts or baskets on Invenio v2.x yet

What about completely disabling/removing these features until somebody requests them? We could later restore them from git and fix them.

**Comment by aw-bib** [comment]
@jirikuncar as of removal of p1,f1 and friends wrt OAI server cf. #2464. Given the mentioned issue the wished for enhancement(!) would favour it.

Concerning dropping alerts and baskets: at join2 we base a lot of workflow on those two functions. Even with new webdeposit workflows and stuff given the way our repos work we need a way to generate _revision lists_ (aka baskets) and send alerts to the editors that there is something to do for them. Probably, there're new and fancy functions available, however, I fear we have quite conservative customers. Ie. they still sort mails in folders instead of tagging them etc.

So from the point of view of our project I'd request alerts and baskets right away. Sorry for the inconveniences. (cc @tiborsimko)

**Comment by kaplun** [comment]
@aw-bib note that in Invenio pu the new holding pen interface (which is a view over running workflows) is being designed precisely to give an administrative view for to dedicated users over tasks/records pending on their action. This could fullfill your needs in a more suitable way?

Can you maybe detail a typical workflow, so that we can see if it can be implemented with all the existing tools of Invenio pu?

**Comment by tiborsimko** [comment]
> So from the point of view of our project I'd request alerts and baskets right away. 

@aw-bib Just to clarify things perhaps, this RFC will not remove alerts/baskets as such, but it will change the search engine API in a way that would "break" alert/basket's use of search, one could say *temporarily* until such time when some Invenio 2.0 installation that uses alerts/baskets will find time to adapt the usage of search engine in those modules.  (Probably as part of the more complete "nextification" of alerts/baskets/tags.)  

Since you are probably not hopping onto the Invenio 2.0 train in the immediate future, this should affect you only much later down the line, and I guess chances are that the CDS team will address these issues by then. CC-ing @kasioumis @egabancho.

> Can you maybe detail a typical workflow, so that we can see if it can be implemented with all the existing tools of Invenio pu?

This is indeed a very good occasion to also look at the new workflow module, which offers powerful curating possibilities, and which could, together with wider usage of tags and notifications, provide a nice alternative to HGF's use of alerts/baskets for curation purposes.

BTW @kaplun please see @aw-bib's presentation at the last Invenio User Group Meeting 2013 on how HGF uses alerts/baskets in curation workflows: http://indico.cern.ch/event/258575/session/0/contribution/28/material/slides/0.pdf

**Comment by egabancho** [comment]
@tiborsimko for alerts we will something with the same functionality for cdslabs (whenever the time comes) but we are planing to drop the use of basket in favor of tags.
But I must say there is a long way for us to reach this point.

**Comment by aw-bib** [comment]
@kaplun @tiborsimko was faster to give you a pointer. :)

> Since you are probably not hopping onto the Invenio 2.0 train in the immediate future

I'm not sure of this, but I agree we are (have to be) very conservative. The more stuff changes the more difficult things will get. OTH we always have to take migration procedures into account, so it could be that this is easier to  migrate to some 2.x than follow up 1.x line and then go for 2.5. Depends on release schedules etc.

> This is indeed a very good occasion to also look at the new workflow module, which offers powerful curating possibilities, and which could, together with wider usage of tags and notifications, provide a nice alternative to HGF's use of alerts/baskets for curation purposes.

:+1: Be sure we'll have a very close look. Every line of our code we can get rid of is a good line. If we manage to move this to deposit we'll not stick to our solution.

However, I'd like to keep a few things in mind:
* We have curators, but it's not like on Inspire that they are full time cataloguers.
* We need to keep the learning curve for them at a minimum.
* We can't call them in at once and give them a decent training.
* A lot of, in your book, old fashioned stuff looks to them like rocket science from mars. I learned that even mail filtering is no established procedure in 2014. Let alone tagging. As Morpheus put it: "Welcome to the desert of the real."

Our experience from the trenches is, that predefining basket and alerts that replenish the baskets and remove stuff that's already worked on (automatically that they never see or change anything there) works well and follows the expectations. We even added some functionality to `alertengine` for the removal part. Note that the improvement in 2014 is, that people don't _print_ the revision lists first. IRL it is still a common workflow that publication lists are sent in by email, printed(!) and then keyed in again. (We work on that...) I even saw hand written lists of URLs(!) handed over the desk for review as common procedure.

So, my feeling is that more modern interfaces like tags, queries  etc. ("what 'all people' use in gmail"; no, IRL they don't) will just not work as the learning curve gets to steep. IOW: I'd have to start to explain what a tag _is_ followed by what a _query_ is followed by how to search for a tag and get back the list. Compared to: (link) is your basket work on this till it's empty.

Still be sure we'll definitely check out deposit and workflows and all the other fancy stuff. ;-)

Sidenote: if you change `perform_request_search()` _API_ into something else this will require a lot of work on our end besides migrations. Most likely we'll need to make up some layer here to keep it at a minimum. I'd prefer if this function stays whatever it calls in the backend. But that's another story.

**Comment by jirikuncar** [comment]
> `perform_request_search()`

@aw-bib I understand your concern and I agree it is *huge* change. Please check discussions in #2478 for record searching and #2662 for formatting. I hope when we are done with refactoring we will be able to provide some wrappers to `1.X` API function and migration scripts.

**Comment by aw-bib** [comment]
@jirikuncar Thanks for the pointer. Keeping sort of a transition API would indeed be nice, especially, as I fear a migration script can not handle our internal and quite complex bean counting tools necessary for evaluations.

**Comment by tiborsimko** [comment]
Note that some other modules need amendments as a result, for example:

```console
$ g g -C 7 -Hni p3 invenio/modules/communities                                             
invenio/modules/communities/models.py-563-        (f1, p1) = self.get_query(provisional=provisional)
invenio/modules/communities/models.py-564-        fields = dict(
invenio/modules/communities/models.py-565-            setName='%s set' % collection_name,
invenio/modules/communities/models.py-566-            setSpec=collection_name,
invenio/modules/communities/models.py-567-            setDescription=self.description,
invenio/modules/communities/models.py-568-            p1=p1, f1=f1, m1='e',
invenio/modules/communities/models.py-569-            p2='', f2='', m2='',
invenio/modules/communities/models.py:570:            p3='', f3='', m3='',
invenio/modules/communities/models.py-571-            setDefinition=''
invenio/modules/communities/models.py-572-        )
invenio/modules/communities/models.py-573-
invenio/modules/communities/models.py-574-        if self.oai_set:
invenio/modules/communities/models.py-575-            update_changed_fields(self.oai_set, fields)
invenio/modules/communities/models.py-576-        else:
invenio/modules/communities/models.py-577-            self.oai_set = OaiREPOSITORY(**fields)
```

**Comment by tiborsimko** [comment]
Note that the work on alerts and baskets (to tags) in the CDS team is planned to start this summer.G (CC @switowski)

We have also agreed upon progressive removal of legacy features in 2.1 (search URLs), 2.2 (more), etc.

Hence this RFC can be closed.  (We'll create separate tickets for individual tasks.)

**Comment by switowski** [comment]
As discussed IRL with @tiborsimko, since the summer student responsible for working on alerts/baskets won't come (I shouldn't have showed him the legacy code base, before he signed the papers :wink:), so those two projects are still available for anybody.
